feat: Add function for detecting Minecraft status effects [skip ci]

WindPerhaps opened the pull request:
Using player_effect(effect) it will take the player's effect and duration, useful for detecting if in vanish (from invisibility), if slowed or blinded, stuff like that. especially useful if using weathered that doesn't detect hpr block or if you just want to use custom hp bars

https://github.com/user-attachments/assets/a3286581-a57a-4d29-ae9c-0e9e95b26202

WindPerhaps committed:
Add function for detecting Minecraft status effects

Using player_effect(effect) it will take the player's effect and duration, useful for detecting if in vanish (from invisibility), if slowed or blinded, stuff like that. especially useful if using weathered that doesn't detect hpr block or if you just want to use custom hp bars

Video example here (just paste into discord): https://cdn.discordapp.com/attachments/796712384592478248/1326703523110457434/2025-01-08_23-56-54_-_Trim.mkv?ex=67806470&is=677f12f0&hm=58676caac03f349a7a74ec800243806f14f7191189c828f9514b4f5bc59de5e1&

Co-Authored-By: James Day <77830514+jambient@users.noreply.github.com>

ryanzhoudev commented on common/src/main/java/com/wynntils/functions/MinecraftFunctions.java:
```diff
@@ -59,4 +64,46 @@ public FunctionArguments.Builder getArgumentsBuilder() {
         }
     }
+
+    public static class PlayerEffectFunction extends Function<Integer> {
```
I think a more descriptive name is EffectDuration or PlayerEffectDuration or something along those lines

WindPerhaps replied:
Maybe we should include minecraft in the name so it doesn't get confused with the wynncraft status effects (StatusEffectFunction()) and StatusEffectActiveFunction())
we could have MinecraftEffectDuration could work but we prolly don't want it getting any longer than that

ryanzhoudev commented on common/src/main/java/com/wynntils/functions/MinecraftFunctions.java:
```diff
@@ -59,4 +64,46 @@ public FunctionArguments.Builder getArgumentsBuilder() {
+            if (McUtils.player().hasEffect(effectHolder)) {
+                MobEffectInstance effectInstance = McUtils.player().getEffect(effectHolder);
+
+                if (effectInstance != null && effectInstance.getDuration() > 0) {
```
Maybe >= 0 so in that 1 second where it displays 0, we can still get a valid number

WindPerhaps replied:
yep, done that now

ryanzhoudev commented on common/src/main/java/com/wynntils/functions/MinecraftFunctions.java:
```diff
@@ -59,4 +64,46 @@ public FunctionArguments.Builder getArgumentsBuilder() {
+        @Override
+        public Integer getValue(FunctionArguments arguments) {
+            String effectName = arguments.getArgument("effectName").getStringValue();
+            ResourceLocation effectLocation = ResourceLocation.withDefaultNamespace(effectName);
```
Instead of doing all these null checks, check out McUtils.player().getActiveEffects(), on each MobEffectInstance you can use .getRegisteredName

WindPerhaps replied:
What would be the need of iterating over all the active effects when it is possible to just lookup using a holder?

ryanzhoudev replied:
Fair, maybe getActiveEffectsMap() instead then. Functions are continuously run, and I feel like that MobEffect registry lookup isn't completely necessary every cycle

WindPerhaps replied:
The key in the active effects map is a Holder to the Mob Effect type which to my understanding can only be acquired from a string using the BuiltInRegistries class which subsequently requires using null checks. However, I did find that the initial null check can be moved into the second one so I have removed that.

ryanzhoudev reviewed (dismissed):
Also missing all localizations, this does not run in devenv at all. Check out en_us.json and the keyPressed function for examples, you can ctrl+f "function.wynntils.keyPressed" to get an idea of what needs to be added

ShadowCat117 commented on common/src/main/java/com/wynntils/core/consumers/functions/FunctionManager.java:
```diff
@@ -555,6 +555,7 @@ private void registerAllFunctions() {
         registerFunction(new MinecraftFunctions.KeyPressedFunction());
         registerFunction(new MinecraftFunctions.MyLocationFunction());
         registerFunction(new MinecraftFunctions.TicksFunction());
+        registerFunction(new MinecraftFunctions.PlayerEffectFunction());
```
Move the register up one line, we keep these alphabetical

ShadowCat117 commented on common/src/main/java/com/wynntils/functions/MinecraftFunctions.java:
```diff
@@ -59,4 +64,46 @@ public FunctionArguments.Builder getArgumentsBuilder() {
+
+            if (effect == null) {
+                return -1; // Effect not found
+            }
```
You can omit the brackets here and just have it all on one line

WindPerhaps replied:
sick ty, that works

ShadowCat117 commented on common/src/main/java/com/wynntils/functions/MinecraftFunctions.java:
```diff
@@ -59,4 +64,46 @@ public FunctionArguments.Builder getArgumentsBuilder() {
+
+            if (effectHolder == null) {
+                return -1; // Effect holder not found
+            }
```
Same here

ShadowCat117 commented on common/src/main/java/com/wynntils/functions/MinecraftFunctions.java:
```diff
@@ -59,4 +64,46 @@ public FunctionArguments.Builder getArgumentsBuilder() {
+        @Override
+        protected List<String> getAliases() {
+            return List.of("player_effect");
+        }
```
Alias is not needed here, the default name is based on the function name `PlayerEffectFunction` with `Function` removed and the rest is snake cased so `player_effect` would be the default name

WindPerhaps committed:
fix: move register

move it to alphabetical order

WindPerhaps committed:
fix: remove null check brackets

simplify if statements

WindPerhaps committed:
fix: add equals check

decimal points considered in duration

WindPerhaps committed:
fix: remove getAliases function

remove unnecessary getAliases function

WindPerhaps committed:
refactor: change function name

change function name from PlayerEffectFunction to MinecraftEffectDuration to avoid confusion

WindPerhaps committed:
fix: remove unnecessary null check

null check did nothing lols

WindPerhaps committed:
fix: put function at end of function name

put function at end of function name

WindPerhaps committed:
docs: localise function

put function into en_us to localise it

magicus commented on common/src/main/java/com/wynntils/functions/MinecraftFunctions.java:
```diff
@@ -59,4 +64,34 @@ public FunctionArguments.Builder getArgumentsBuilder() {
+        @Override
+        public Integer getValue(FunctionArguments arguments) {
+            String effectName = arguments.getArgument("effectName").getStringValue();
+            ResourceLocation effectLocation = ResourceLocation.withDefaultNamespace(effectName);
```
This is starting to reach the level of complexity that is "allowed" in a function, before you need to move the logic to a model or other component. I think this is still fine since it is "just" about converting an argument to an internal java type, but it is a borderline case.

magicus commented:
I'm a bit surprised that this looks at the *Minecraft* effect. What we would like to see is the *Wynncraft* effects, like "Vanish" rather than "invisibility". I thought that the CharacterModel held an up-to-date list of applied status effects? It seems to me that this is what we should be checking instead.

ryanzhoudev commented:
> I'm a bit surprised that this looks at the _Minecraft_ effect. What we would like to see is the _Wynncraft_ effects, like "Vanish" rather than "invisibility". I thought that the CharacterModel held an up-to-date list of applied status effects? It seems to me that this is what we should be checking instead.

We already have this to some extent, `status_effect_active` that returns booleans based on if specified Wynn effective is active, and `status_effects` to return the entire list.

WindPerhaps commented:
> I'm a bit surprised that this looks at the _Minecraft_ effect. What we would like to see is the _Wynncraft_ effects, like "Vanish" rather than "invisibility". I thought that the CharacterModel held an up-to-date list of applied status effects? It seems to me that this is what we should be checking instead.

yeah you guys do already have access to check Wynncraft effects and I did actually make the original vanish bar using it but for some reason Wynncraft's effect are like delayed before they're listed, so it would take a second to tell me if in in vanish especially if I had just casted a couple spells which is enough to mess up my cycle or even get me killed. The Minecraft effect appears instantly and consistently so, alongside being useful for stuff like checking your slowness duration and stuff like that with less delay and better accuracy

I cant tell if I 100% understood what you said due to my little knowledge on coding and kinda the human language as a whole but hopefully that answers your question

magicus reviewed (approved):
Fair enough. I see that this is properly named "minecraft effect", so I guess that makes it unambiguous to what it does. The code itself look fine. Thank you @WindPerhaps for contributing to Wynntils!

magicus committed:
Merge branch 'main' into main
